snap/snapcraft.yaml

```diff
@@ -303,137 +227,84 @@ parts:
-            LDFLAGS="-L${SNAPCRAFT_STAGE}/libuv/.libs -L${SNAPCRAFT_STAGE}/liblz4/lib -L${SNAPCRAFT_STAGE}/libnsl/src" \
-            UV_CFLAGS="-I${SNAPCRAFT_STAGE}/libuv/include" \
-            UV_LIBS="-L${SNAPCRAFT_STAGE}/libuv/.libs" \
-            LZ4_CFLAGS="-I${SNAPCRAFT_STAGE}/liblz4/lib" \
-            LZ4_LIBS="-L${SNAPCRAFT_STAGE}/liblz4/lib" \
-            SQLITE_CFLAGS="-I${SNAPCRAFT_STAGE}/libsqlite3" \
-            ./configure --disable-shared --enable-build-raft
-      make
-
-      mkdir -p $SNAPCRAFT_PART_INSTALL/libdqlite
-      cp -r $SNAPCRAFT_PART_BUILD/.libs $SNAPCRAFT_PART_INSTALL/libdqlite/.libs
-
-      mkdir -p $SNAPCRAFT_PART_INSTALL/libs
-      cp -r $SNAPCRAFT_PART_BUILD/.libs/*.a $SNAPCRAFT_PART_INSTALL/libs
-
-      mkdir -p $SNAPCRAFT_PART_INSTALL/libs/include
-      cp -r $SNAPCRAFT_PART_BUILD/include/* $SNAPCRAFT_PART_INSTALL/libs/include
-    stage:
-    - -*.*
+      ./configure --disable-shared --enable-build-raft --prefix "${CRAFT_STAGE}/snap-linux-musl"
+
+      make install
 
   jujud:
     after:
       - libdqlite
-    plugin: juju-go
-    go-channel: 1.22/stable
+    plugin: nil
     source: .
-    go-packages:
-      - github.com/juju/juju/cmd/jujud-controller
-    go-external-strings:
-      github.com/juju/juju/core/version.GitCommit: ""
-      github.com/juju/juju/core/version.GitTreeState: ""
-      github.com/juju/juju/core/version.build: ""
-    go-static: true
-    go-strip: true
-    go-buildtags: ["libsqlite3", "dqlite", "notest"]
-    go-cgo-enabled: "1"
-    go-cgo-cc: "musl-gcc"
-    go-cgo-cflags: "-I/usr/local/musl/include"
-    go-cgo-ldflags: "-luv -ldqlite -llz4 -lsqlite3 -Wl,-z,stack-size=1048576"
-    go-cgo-ldflags-allow: "(-Wl,-wrap,pthread_create)|(-Wl,-z,now)"
+    build-snaps:
+      - go/1.22/stable
+    build-packages:
+      - musl-tools
     override-build: |
-      snapcraftctl build
-      mv ${SNAPCRAFT_PART_INSTALL}/bin/jujud-controller ${SNAPCRAFT_PART_INSTALL}/bin/jujud
+      set -ex
+
+      # TODO: use make plugin.
+      export CC=musl-gcc
+      export PKG_CONFIG_PATH="${CRAFT_STAGE}/snap-linux-musl/lib/pkgconfig"
+      export PACKAGES="libuv dqlite liblz4 sqlite3"
+      export GOBIN="${CRAFT_PART_INSTALL}/bin"
+      export CGO_ENABLED=1
+      export CGO_CFLAGS="$(pkg-config --cflags ${PACKAGES})"
+      export CGO_LDFLAGS="$(pkg-config --libs ${PACKAGES}) -Wl,-z,stack-size=1048576"
+      export CGO_LDFLAGS_ALLOW="(-Wl,-wrap,pthread_create)|(-Wl,-z,now)"
+      export GO_LDFLAGS='-s -w -extldflags "-static" -linkmode "external" -X github.com/juju/juju/version.GitCommit= -X github.com/juju/juju/version.GitTreeState= -X github.com/juju/juju/version.build='
+      go mod download
+      go install -tags libsqlite3,dqlite -ldflags "${GO_LDFLAGS}" github.com/juju/juju/cmd/jujud
```

Comment:
This needs to change.

```
go install -tags libsqlite3,dqlite -ldflags "${GO_LDFLAGS}" github.com/juju/juju/cmd/jujud-controller
mv ${SNAPCRAFT_PART_INSTALL}/bin/jujud-controller ${SNAPCRAFT_PART_INSTALL}/bin/jujud
```

Reply:
This has fixed the test error, thanks a lot @SimonRichardson !